IG updates pre-ballot (#20)
HL7/smart-web-messaging: fsh/ig-data/input/pagecontent/index.md

const message = {
  "messagingHandle": "<smart_web_messaging_handle> from SMART launch context",
  "messageId":       "<some new guid>",
  "messageType":     "scratchpad.create",
  "payload":         {}  // See below.
};

targetWindow.postMessage(message, targetOrigin);
```

In the EHR, the message is received and is handled as shown below.

```js
window.addEventListener("message", function(event) {
  if (event.origin !== "<the app's expected origin>") {
    return;  // Ignore unknown origins.
  }

  //
  // TODO: Handle the message here by using the contents of event.data.
  //

  // Send a response back to the app.
  const response = {
    ... // See below for more details on the response properties.
  };
  event.source.postMessage(response, event.origin);
});
```

#### Response Properties
The EHR SHALL send a response message with the following properties:

| Property              | Optionality  | Type   | Description |
| --------------------- | ------------ | ------ | ----------- |
| `messageId`           | REQUIRED     | ID     | A unique ID for this message generated by the application. |

> | `responseToMessageId` | CONDITIONAL  | ID     | The `messageId` of the received message that this message is in response to.  MAY NOT exist in [unsolicited response messages](#unsolicited-responses) beyond the first. |

Reviewer: ```suggestion
| `responseToMessageId` | REQUIRED  | ID     | The `messageId` of the received message that this message is in response to.|
```

| `payload`             | REQUIRED     | object | The message content as specified by the `messageType` of the request message.  See [below](#response-payload). |
{:.grid}

> 

Reviewer: ```suggestion

In addition to sending Web Messaging Responses to the app, the EHR MAY also send "unsolicited" requests to the app, omitting  a `responseToMessageId` property. The semantics of such requests are are implementation-specific.
```

#### Response Payload
The response message `payload` properties will vary based on the request `messageType`.  See message types below for details.

#### Response Target Origin
It is assumed that the EHR already knows the proper `targetOrigin` to use in its
call to [`window.postMessage`] because it has demonstrated that it can SMART
launch the app at a known SMART launch URL.  The SMART launch URL SHALL be
prefixed with the proper value to use for the app `targetOrigin`, and the launch
URL SHALL not redirect users to a different origin after launch.

#### Detailed Example Response
In a more detailed example response, the EHR may send one return message like:

```js
window.addEventListener("message", function(event) {
  if (event.origin != "<the app's expected origin>") {
    return;  // Ignore unknown origins.
  }

  const messageType = event.data.messageType;

  // The response payload is modeled after Bundle.entry.response.
  // See: https://www.hl7.org/fhir/bundle-definitions.html#Bundle.entry.response

>   var location = "Example/123";

Reviewer: Can we inline these below? The variables aren't used elsewhere (and if we do keep declarations, should be `const`).

Author: Yeah, I was trying to suggest that the `TODO: handle message here` code below the declarations might update those variables, but I can see that it's a bit awkward to do it that way.  Inline is clearer.

```js
targetWindow.postMessage({
  "messageId": "<some new guid>",
  "messageType": "ui.done",
  "payload": {
    "activityType": "problem-add",
    "activityParameters": {
      // Each ui activity defines its optional and required params.  See the
      // Activity Catalog for more details.
      "problem": {
        "resourceType": "Condition",
        "patient": "123",
      }
    }
  }
}, targetOrigin);
```

Similarly, the SMART app can use the `ui.LaunchActivity` message type to request
navigation to a different activity *without* closing the app:

```js
targetWindow.postMessage({
  "messageId": "<some new guid>",
  "messageType": "ui.launchActivity",
  "payload": {
    "activityType": "problem-add",
    "activityParameters": {
      // Each ui activity defines its optional and required params.  See the
      // Activity Catalog for more details.
      "problem": {
        "resourceType": "Condition",
        "patient": "123",
      }
    }
  }
}, targetOrigin);
```

The EHR SHALL respond to all `ui` message types with a payload that includes a
boolean `success` parameter and an optional `details` string:

```js

> event.source.postMessage({

Reviewer: This is missing the listener code from the examples above -- so `event` isn't clearly defines, and it's not clear whether event.origin has been checked and is safe to postMessage to).

Author: Fair enough.  I thought about adding the event listener code back in, but it muddled the example.

messages, i.e., messages with no `responseToMessageId`, after a client's initial
request.

> ### Influence the EHR UI: `messageType ui.*`

Reviewer: The string `messageType ui` won't appear in code every, so I think it's clearer if we don't put that string in backticks.

Author: OK, I can fix that.

{:.grid}

The `activityType` property conveys an activity type drawn from the SMART Web
Messaging [Activity Catalog]. In general, these activities follow the same
naming conventions as entries inthe CDS Hooks catalog (`noun-verb`), and will
align with CDS Hooks catalog entries where feasible. The `activityType` property
conveys a navigation target such as `problem-add` or `order-sign`, indicating
where EHR should go to after the ui message has been handled. An activity may
specify additional parameters that can be included in the call as additional
properties.

The `activityParameters` property conveys parameters specific to an activity
type. See the SMART Web Messaging [Activity Catalog] for details.

*Note:* A SMART app launched in the context of CDS Hooks should generally not
need to specify an `activityType` or `activityParameters` with the `ui.done`
message, because the EHR tracks the context in which the app was launched (e.g.,
order entry) and can navigate to an appropriate follow-up screen based on this
context.

An example of a `ui.done` message from an app to the EHR is shown below:

```js

> window.postMessage({

Reviewer: `window.postMessage` would mean sending a message to one's own window, because in the browser environment `window` is bound to the window holding the current DOM... which isn't what we want to demonstrate here.

Author: Ah, thanks for that.  I noticed that in the examples we were talking about having the app send messages to the EHR, but we were using `appWindow` as the variable name and that seemed backwards.  To match earlier examples, should the app examples continue using `targetWindow` as the variable name?

Author: I replaced all the `window` references in "app" code examples with `targetWindow` (resembling the original example).  I also replaced all `window.postMessage(..., clientAppOrigin)` references with `event.source.postMessage(..., event.origin)` for the EHR code examples.

}, targetOrigin);
```

As described above, the EHR responds to all `scratchpad` message types with a
payload that matches FHIR's [`Bundle.entry.response`] data model.  For instance,
the response to a `scratchpad.create` that adds a new prescription to the
scratchpad (and assigns id `456` to this draft resource) might look like:

```js

> event.source.postMessage({

Reviewer: Same comment on context here for event

Author: Got it.
